simplified/HordeAd/Core/Ast.hs

```diff
@@ -31,19 +38,138 @@ import           GHC.TypeLits (KnownNat, Nat, type (+))
 import           Numeric.LinearAlgebra (Numeric)
 import           System.IO.Unsafe (unsafePerformIO)
 import           Text.Show.Functions ()
+import           Unsafe.Coerce (unsafeCoerce)
 
 import HordeAd.Internal.OrthotopeOrphanInstances ()
 
--- * Definitions
+-- * GHC.Nat-indexed lists, originally by Tom Smeding
+
+-- This needs to be moved to a better place, probably its own module,
+-- once its clear which other modules depend on it.
+
+-- | An index in an n-dimensional array. The fastest-moving index is at the
+-- last position; thus the index 'Z :. i :. j' represents 'a[i][j]' in
+-- traditional C notation.
+data Index (n :: Nat) i where
+  Z :: Index 0 i
+  S :: i -> Index n i -> Index (n + 1) i
+
+deriving stock instance Functor (Index n)
+
+instance Show i => Show (Index n i) where
+  showsPrec _ Z = showString "Z"
+  showsPrec d (idx :. i) = showParen (d > 3) $
+    showsPrec 3 idx . showString " :. " . showsPrec 4 i
+
+-- I'm afraid, I can't do the unsafeCoerces below with this order
+-- of argument in :., can I? So I need this instead:
+pattern (:.) :: forall n1 i. forall n. n1 ~ (n + 1)
+             => Index n i -> i -> Index n1 i
+pattern (:.) is i = S i is
+{-# COMPLETE Z, (:.) #-}
+infixl 3 :.
+-- This fixity is stolen from Accelerate:
+--   https://hackage.haskell.org/package/accelerate-1.3.0.0/docs/Data-Array-Accelerate.html#t::.
+
+tailIndex :: Index (1 + n) i -> Index n i
+tailIndex Z = error "tailIndex: impossible pattern needlessly required"
+tailIndex (is :. _i) = is
+
+takeIndex :: forall len n i. KnownNat len
+          => Index (len + n) i -> Index n i
+takeIndex ix = unsafeCoerce $ take (valueOf @len) $ unsafeCoerce ix
+
+dropIndex :: forall len n i. KnownNat len
+          => Index (len + n) i -> Index n i
+dropIndex ix = unsafeCoerce $ drop (valueOf @len) $ unsafeCoerce ix
+
+permutePrefixIndex :: [Int] -> Index n Int -> Index n Int
+permutePrefixIndex p ix =
+  let l = unsafeCoerce ix
+  in (unsafeCoerce :: [Int] -> Index n Int)
+     $ V.toList $ VS.fromList l V.// zip p l
```

Comment:
Please check that `p` has the right length, otherwise this is wildly unsafe (the `zip` will stop at the end of the shortest list)

Reply:
Well spotted.

Reply:
Fixed.